### PR TITLE
optimized GetAvailableUpdates in AddonManager

### PR DESCRIPTION
**Pull request by burlizzi**
## Description

optimized a little (factor of 10) detection of update need for plugins

## Motivation and Context

every time the plugin directory needs to be updated (e.g. after video plugin navigation) the directory takes up to 10 seconds to find out  just the availability of new updates in the addon database on my RPi 3

## How Has This Been Tested?

on RPi3, (Libreelec 9.0.2) with enough addons installed (70 in my case) navigate into any Video plugin directory (Youtube) then exit and move up to movies, it will freeze the UI for about 10 sec,

2019-05-18 11:40:30.160 T:1454371696   DEBUG: CAddonMgr::GetAvailableUpdates took 10224 ms

 with the change it is reduced to 1 sec

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/AddonManager.cpp [hunk 1]
```diff
@@ -448,12 +448,22 @@ VECADDONS CAddonMgr::GetAvailableUpdates()
 
   VECADDONS updates;
   VECADDONS installed;
+  VECADDONS all_addons;
   GetAddons(installed);
+  m_database.GetRepositoryContent(all_addons);
+  std::map<std::string, AddonPtr> last_versions;
+  for (const auto& addon : all_addons)
+  {
+	  const auto& latest_known = last_versions.find(addon->ID());
```
Comment (Rechi) [inline xbmc/addons/AddonManager.cpp]: wrong indentation, correct is two spaces and never use tabs

File: xbmc/addons/AddonManager.cpp [hunk 2]
```diff
@@ -448,12 +448,22 @@ VECADDONS CAddonMgr::GetAvailableUpdates()
...
+	  if (latest_known == last_versions.end() || addon->Version() > latest_known->second->Version())
```
Comment (Rechi) [inline xbmc/addons/AddonManager.cpp]: wrong indentation

File: xbmc/addons/AddonManager.cpp [hunk 3]
```diff
@@ -448,12 +448,22 @@ VECADDONS CAddonMgr::GetAvailableUpdates()
...
+		  last_versions[addon->ID()]=std::move(addon);
```
Comment (Rechi) [inline xbmc/addons/AddonManager.cpp]: wrong indentation and spaces before and after `=` are missing

File: xbmc/addons/AddonManager.cpp [hunk 4]
```diff
@@ -448,12 +448,22 @@ VECADDONS CAddonMgr::GetAvailableUpdates()
...
+  }
+
   for (const auto& addon : installed)
   {
-    AddonPtr remote;
-    if (m_database.GetAddon(addon->ID(), remote) && remote->Version() > addon->Version())
-      updates.emplace_back(std::move(remote));
+	const auto& remote= last_versions.find(addon->ID());
```
Comment (Rechi) [inline xbmc/addons/AddonManager.cpp]: wrong indentation and space before `=` is missing

File: xbmc/addons/AddonManager.cpp [hunk 5]
```diff
@@ -448,12 +448,22 @@ VECADDONS CAddonMgr::GetAvailableUpdates()
...
+    if (remote!= last_versions.end() && remote->second->Version() > addon->Version())
```
Comment (Rechi) [inline xbmc/addons/AddonManager.cpp]: wrong indentation and space before `!=` is missing

**Comment by ksooo** [comment]
Please squash the two commits down to one and force-push.

File: xbmc/addons/AddonManager.cpp [hunk 6]
```diff
@@ -448,12 +448,22 @@ VECADDONS CAddonMgr::GetAvailableUpdates()
...
   GetAddons(installed);
+  m_database.GetRepositoryContent(all_addons);
+  std::map<std::string, AddonPtr> last_versions;
+  for (const auto& addon : all_addons)
+  {
+    const auto& latest_known = last_versions.find(addon->ID());
+    if (latest_known == last_versions.end() || addon->Version() > latest_known->second->Version())
+      last_versions[addon->ID()] = std::move(addon);
+  }
+
```
Comment (ksooo) [inline xbmc/addons/AddonManager.cpp]: `installed` is used first time after the new loop. I would place `GetAddons(installed);` directly before the loop it is first used (like it was before).

**Comment by pkerling** [comment]
Wouldn't it make more sense to investigate why `m_database.GetAddon` is (supposedly) so slow and optimize there?

**Comment by ksooo** [comment]
> Wouldn't it make more sense to investigate why m_database.GetAddon is (supposedly) so slow and optimize there?

 Makes sense, not instead of this PR, but additionally, I'd say.

**Comment by pkerling** [comment]
> Makes sense, not instead of this PR, but additionally, I'd say.

If you do a map lookup in `CAddoonDatabase` instead, it would make this PR completely obsolete (if that is easily possible)

**Comment by burlizzi** [comment]
m_database.GetAddon is not that much slow, it  just call SQL engine which is definitely slower than a std map search. As it is done in a loop the inefficiency is amplified.
the improvement here is to make a single SQL call at the beginning and iterate a local object

**Comment by burlizzi** [comment]
Database is in the name of CAddonDatabase, map lookup seems a violation of the design. 
I could think of some kind of caching but GetAddon is an atomic call, I wouldn't know how/when to populate the map cache without impacting on the existing code.

**Pull request merged**